code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java

```diff
@@ -239,6 +243,60 @@ public void setVIP(Service service) {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    private List<PortSpec> getServicePorts(Service service) {
+        Map<String, Object> launchConfigData = DataAccessor.fields(service)
+                .withKey(ServiceDiscoveryConstants.FIELD_LAUNCH_CONFIG).withDefault(Collections.EMPTY_MAP)
+                .as(Map.class);
+        if (launchConfigData.get(InstanceConstants.FIELD_PORTS) == null) {
+            return new ArrayList<>();
+        }
+        List<String> specs = (List<String>) launchConfigData.get(InstanceConstants.FIELD_PORTS);
+        List<PortSpec> ports = new ArrayList<>();
+        for (String spec : specs) {
+            ports.add(new PortSpec(spec));
+        }
+        return ports;
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void setPorts(Service service) {
+        List<PortSpec> ports = getServicePorts(service);
+        Map<String, Object> launchConfigData = DataAccessor.fields(service)
+                .withKey(ServiceDiscoveryConstants.FIELD_LAUNCH_CONFIG).withDefault(Collections.EMPTY_MAP)
+                .as(Map.class);
+        Account env = objectManager.loadResource(Account.class, service.getAccountId());
+        List<String> newPorts = new ArrayList<>();
+        for (PortSpec port : ports) {
+            if (port.getPublicPort() == null) {
+                PooledResource resource = poolManager.allocateOneResource(env, service,
```

Comment:
I don't think this will work if the service has multiple random ports.  The allocateOneResource will return the same port each time.  Can you write a test that has a service with more that one random port that needs to be assigned.